src/cisco_gnmi/flatten.py

```diff
@@ -0,0 +1,204 @@
+"""
+import json
+import logging
+from . import proto
+
+
+def flatten(message, convert_strings=True, ignore_delete=True, origin_as_module=False):
+    """Top level convenience function which accepts a protobuf
+    message and indirects to the appropriate handler function to
+    flatten the data to a simple, flat xpath:value dict representation.
+    Currently only handles GetResponse and SubscribeResponse.
+    """
+    flatten_method_map = {
+        proto.gnmi_pb2.GetResponse: _flatten_get_response,
+        proto.gnmi_pb2.SubscribeResponse: _flatten_subscribe_response,
+    }
+    identified = False
+    flattened_message = None
+    for message_class, flatten_method in flatten_method_map.items():
+        if isinstance(message, message_class):
+            identified = True
+            flattened_message = flatten_method(
+                message, convert_strings, ignore_delete, origin_as_module
+            )
+    if not identified:
+        raise Exception("Flatten not yet supported for message class!")
+    return flattened_message
+
+
+def flatten_yang_json(prefix, yang_json, convert_strings=True, implicit_keys=True):
+    """Flattens a JSON structure with special consideration
+    around strings - inlining like keys and attempting integer
+    conversion given the uint64-as-string JSON encoding rules.
+    Recursive implementation, many dict updates, not intensely validated.
+    """
+    flattened = {}
+    keys = {}
+    values = {}
+    to_traverse = {}
+    if isinstance(yang_json, list):
+        for elem in yang_json:
+            flattened.update(flatten_yang_json(prefix, elem))
+    elif isinstance(yang_json, dict):
+        for key, value in yang_json.items():
+            if isinstance(value, str):
+                potential_key = False
+                if convert_strings:
+                    try:
+                        values[key] = int(value)
+                    except ValueError:
+                        try:
+                            values[key] = float(value)
+                        except ValueError:
+                            potential_key = True
+                else:
+                    potential_key = True
+                if potential_key and implicit_keys:
+                    keys[key] = value
+                else:
+                    values[key] = value
+            # Fun fact, boolean is a sub of int in Python
+            elif isinstance(value, bool) and implicit_keys:
+                keys[key] = json.dumps(value)
+            elif isinstance(value, (int, float)):
+                values[key] = value
+            elif isinstance(value, (dict, list)):
+                to_traverse[key] = value
+            else:
+                raise Exception("Unhandled element type!")
+        key_string = ""
+        for key, value in keys.items():
+            key_string += "[{key}={value}]".format(key=key, value=value)
```

Comment:
Values to keys may have illegal url characters so you should put them in quotes (I think that is good enough)
``"[{key}=\"{value}\"]"``